netx/measurable/measurable.go

```diff
@@ -0,0 +1,219 @@
+// Package measurable makes DNS, connect, TLS, and HTTP measurable.
+package measurable
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+	"net/http"
+	"time"
+)
+
+// Operations contains measurable operations
+type Operations interface {
+	LookupHost(ctx context.Context, domain string) ([]string, error)
+	DialContext(ctx context.Context, network, address string) (net.Conn, error)
+	Handshake(ctx context.Context, conn net.Conn, config *tls.Config) (
+		net.Conn, tls.ConnectionState, error)
+	RoundTrip(req *http.Request) (*http.Response, error)
+}
+
+type operationsKey struct{}
+
+// WithOperations returns a new context using the specified measurable operations.
+func WithOperations(ctx context.Context, operations Operations) context.Context {
+	return context.WithValue(ctx, operationsKey{}, operations)
+}
+
+// ContextOperations returns the Operations associated with the context.
+func ContextOperations(ctx context.Context) Operations {
+	operations, _ := ctx.Value(operationsKey{}).(Operations)
+	return operations
+}
```

Comment:
This design allows us to change basically everything with every request using a context. This is nice but at the same time it becomes very confusing to understand what is happening when you are reading the code. It seems that we should not pursue this strategy and rather see to use the context as sparingly as possible.